### ISSUE TITLE
MathJax Documentation Is Ambiguous on How Environments Should Be Used

### ISSUE THREAD

**Issue by ickc** [mathjax/MathJax-docs#146]
In the end of [MathJax TeX and LaTeX Support — MathJax 2.6 documentation](http://docs.mathjax.org/en/latest/tex.html), it is mentioned under heading [Environments](http://docs.mathjax.org/en/latest/tex.html#environments) that some LaTeX environments is replicated in MathJax. However it didn't specify clearly how to use it.

In particular, in the same page under heading [Defining TeX macros](http://docs.mathjax.org/en/latest/tex.html#defining-tex-macros):

	Unlike actual TeX, however, in order for MathJax to process these, they must be enclosed in math delimiters (since MathJax only processes macros in math-mode).

Although that was from a different context, but it still might leads one to think the environments should also be enclosed in math delimiters, e.g. `\[\begin{equation}...` (although in LaTeX `\begin{equation}...` should not be used with `\[...`). Furthermore, MathJax is pretty lenient on giving you the correct result even if you write "wrong" code.

The result of this is different writers will interpret it differently and so write codes differently. And hence authors of different Markdown converter (mmd and pandoc are of interests here) might made different assumptions on the source code in Markdown where environment is involved and to be converted in HTML and LaTeX.

I opened a GitHub Pages in [Testing LaTeX Environments Usage in MathJax From Markdown Convertion](https://ickc.github.io/mathjax-latex-md-mmd-pandoc/) (and the GitHub repository in [ickc/mathjax-latex-md-mmd-pandoc: Testing LaTeX Environments Usage in MathJax From Markdown Convertion (including mmd and pandoc)](https://github.com/ickc/mathjax-latex-md-mmd-pandoc)) to discuss the issue and the consequence of it in terms of Markdown to HTML generations, and provided some test files.

On MathJax's side, one thing we could do to improve the situation is probably to clarify that while macro definition has to be in math delimiters, environments do not. And probably add more test files in the MathJax repository.

**Comment by dpvc** [comment]
In early versions of MathJax, you did need to use math delimiters around the environments.  That is why some early adopters (like pandoc) use them.  That changed (don't remember the version, but it was early on), and the early adopters didn't always update their approach.  Using math delimiters around the environments doesn't hurt anything (from MathJax's point of view), and the results are the same, so they didn't feel the need to change their code.

Because the behavior changed, MathJax provides a configuration parameter to get the old behavior.  See the [tex2jax documentation](http://docs.mathjax.org/en/latest/options/tex2jax.html?highlight=processenvironments) for the `processEnvironments` option.  There are a few other things outside of math mode that MathJax can also process (e.g., `\ref{}` and `\eqref{}`, and `\$`), also controlled by configuration options.

It is true that the documentation could be better about this.

**Comment by ickc** [comment]
Thanks and sorry I didn't see the doc on tex2jax.

And from my test pandoc do not requires additional delimiter, but mmd does.

The main problem (and the main reason I dig into this) is for single markdown source to generate both correct HTML (with MathJax) and LaTeX code, that I explained in details in the gh-pages above.

The main problems is that all 3 (MathJax, pandoc, mmd) doesn't document how the environments (and other things like (/ref) you mentioned should be used.

As you mentioned, MathJax did document the behavior in the tex2jax doc, but it is provided as options. I understand MathJax's intention to provide options is good, for both compatibility and for fulfilling different audiences' usage. But I think probably MathJax should also give a recommended usages for all of them (as far as I can see, for the features it supports, it can take the LaTeX as it is without adding delimiters. So probably this is a good and very simple way to recommend it).

And I understand it's not the job of MathJax on how markdown converters should work. If they are to change anything to follow MathJax newer convention they would break backward compatibility. But I'm not going to mention it here.

In short, what I am asking is a better documentation in the sense that although different options are provided, but a recommended standard way should be suggested and documented.

**Comment by pkra** [comment]
> In short, what I am asking is a better documentation in the sense that although different options are provided, a recommended standard way should be suggested and documented.

If there ever was a movement similar to commonmark to standardize LaTeX-like syntax for webby situations (such as markdown), we'd be happy to contribute.

**Comment by ickc** [comment]
To be clear I am not asking MathJax to standardize the web, nor any Markdown converter, but to provide an recommended way out of many different ways to write LaTeX environments in MathJax. Just like the MathJax has many configurable options, but there is a recommendation on how one can start. In terms of LaTeX environments, there isn't such recommendation in the documentation.

**Comment by pkra** [comment]
> Just like the MathJax has many configurable options, but there is a recommendation on how one can start.

Our default settings for various configuration options do that. Beyond it, we don't really want to tell people how to write their documents. For example, we also don't tell people to use `\mathbf` over `\bf` even though in real LaTeX you really should. Again, we're happy to highlight style guides that help other projects and even provide help with configuring MathJax to enforce a certain style.

But I would suggest to instead start a discussion with the maintainers of markdown converters with math support.

**Comment by ickc** [comment]
Already did with the mmd and discouraged. So I am giving up to talk to anyone in pandoc. Thanks anyway.

**Comment by pkra** [comment]
Sorry to hear that. Does mmd actually support TeX-like math? I know kramdown does and some pagedown flavors like stackedit do. 

I guess an alternative would be to bring math-mode-TeX-to-something converters together but that's probably even harder.

**Comment by ickc** [comment]
The situation of mmd is more complicated. pandoc basically works (almost) perfectly in html/tex generation if the input code is identical to what latex expects (because MathJax supports so).

But for mmd, it expects an extra math delimiter. 
e.g. `\\[\begin{equation}` becomes `\[\begin{equation}` in HTML (where MathJax understood), and becomes `\begin{equation}` in LaTeX (correct).
In the case of `\\[\begin{aligned}`, the html generation becomes `\[\begin{aligned}` (correct), but the LaTeX generation becomes `\begin{aligned}` (leads to error).
Basically because it always expects an extra math delimiter, when an extra math delimiter is really needed (e.g. `aligned`), it will be trimmed out.

And when I say discouraged, I mean the suggestion on improving the documentation is discouraged. There is actually a workaround in the above situation, that if an extra line break is inserted between `\\[` and `\begin`, it won't be trimmed out.

Going back to my original point, I think the cleanest way to aim at both HTML and TeX generation will be following exactly the way LaTeX needed, where MathJax understands, and pandoc will output valid HTML and TeX (because even if pandoc don't recognize `\begin...` as math, it will pass it as raw html and MathJax will pick it up).

So if the documentation shouldn't

> ...tell people how to write their documents...

may be more test suite can be given in [MathJax/test at master · mathjax/MathJax](https://github.com/mathjax/MathJax/tree/master/test)? It is from there I actually found out in MathJax `\begin{...` can be used without delimiter.

While I am brainstorming, I am thinking the current MathJax documentation is like a big user manual. Might be a quick guide can be offered in parallel with that. Sorry to say that I didn't read *through* the MathJax documentation from beginning to the end. I only read sections from sections that seemed relevant. And I never thought the section on "tex2jax" would have the information I was looking for (I always thought it should be in "MathJax TeX and LaTeX Support").

**Comment by ickc** [comment]
I have another suggestion about how the documentation can be improved:

May be there can be a section collecting all the situations that MathJax *requires* a different way of writing the math codes from the LaTeX counter parts.

If put in this way, we are not telling how people should write their code, but to realize the cases where there is a *mandatory differences between MathJax and LaTeX math*. One example is what is mentioned already in [Defining TeX macros](http://docs.mathjax.org/en/latest/tex.html#defining-tex-macros), but I don't know if there are others. Say if this is the only instance, then it could be mentioned there this is the only exception when you *need* a different writing from LaTeX's. If that is not the only case an extra delimiter is required, then putting them altogether can make it much more clear on how something *has to be written* in MathJax.

**Comment by pkra** [comment]
> May be there can be a section collecting all the situations that MathJax requires a different way of writing the math codes from the LaTeX counter parts.

That's probably #26.
